core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,307 @@
+  }
+
+  /**
+   * Method to set file tracking information to temporary file to read.
+   * This is to read / write file tracking information from / to XMLInputFormat and XMLRecordReader using temporary
+   * file, as there is no direct way to read / write information from / to dataset.
+   * @param context
+   * @param conf
+   * @throws IOException
+   */
+  private void setFileTrackingInfo(BatchSourceContext context, Configuration conf) throws IOException {
+    String tableName = this.config.tableName;
+    if (StringUtils.isNotEmpty(tableName)) {
+      try {
+        processedFileTrackingTable = context.getDataset(tableName);
+        if (processedFileTrackingTable != null) {
+          File tempFile = new File(tableName);
+          // if file exists, then delete it and create new.
+          if (tempFile.exists()) {
+            tempFile.delete();
+          }
+          tempFile.createNewFile();
+          FileWriter fw = new FileWriter(tempFile);
+          BufferedWriter bw = new BufferedWriter(fw);
+          CloseableIterator<KeyValue<byte[], byte[]>> iterator = processedFileTrackingTable.scan(null, null);
+
+          //write file tracking information to temporary file.
+          while (iterator.hasNext()) {
+            KeyValue<byte[], byte[]> keyValue = iterator.next();
+            bw.write(new String(keyValue.getKey(), Charsets.UTF_8) + "\n");
+          }
+          bw.close();
+          conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_DATA_TEMP_FILE, tableName);
+        }
+      } catch (IOException ioException) {
+        throw ioException;
+      } catch (DatasetInstantiationException exception) {
+        throw exception;
+      }
+    }
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Object> input, Emitter<StructuredRecord> emitter) throws Exception {
+    Map<String, String> xmlRecord = (Map<String, String>) input.getValue();
+    String fileName = "";
+    String record = "";
+    for (String key : xmlRecord.keySet()) {
+      fileName = key;
+      record = xmlRecord.get(key);
+    }
+
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_XML_SCHEMA)
+      .set("offset", input.getKey().get())
+      .set("fileName", fileName)
+      .set("record", record)
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    super.onRunFinish(succeeded, context);
+    if (StringUtils.isNotEmpty(this.config.tableName)) {
+      try {
+        //read file tracking information updated by XMLRecordReader and put it in dataset.
+        File file = new File(this.config.tableName);
+        BufferedReader reader = new BufferedReader(new FileReader(file));
+        String line = null;
+        while ((line = reader.readLine()) != null) {
+          long processingTime = new Date().getTime();
+          processedFileTrackingTable.write(line.getBytes(Charsets.UTF_8),
```

Comment:
In KeyValue table only processed date should get updated for existing key (i.e. file name). So table should grow for newly processed file only. Shall we clear table data before some particular time period like before 30 day. Please let me know your inputs.